File: Tree/tree_DFS.cpp
/*
Tree DFS : visited array 대신 parent array && depth array 사용
*/
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[10];
int p[10];          // 각 정점의 부모 번호를 저장할 P 배열
int depth[10];

void DFS(int root){
    stack<int> S;
    S.push(root);
    while (!S.empty()){
        int cur = S.top(); S.pop();
        cout << cur << ' ';
        for (int nxt : adj[cur]){
            if (p[cur] == nxt) continue;    // nxt가 cur의 부모 확인 -> 부모인 인접값은 제외 && 자식 인접값만 추가
            S.push(nxt);
            p[nxt] = cur;                   // nxt의 부모로 cur를 저장 
            depth[nxt] = depth[cur] + 1;    // Tree depth ++
        }
    }
}

// 재귀 DFS 
void DFS_2(int cur){
    cout << cur << ' ';
    for (int nxt : adj[cur]){
        if (p[cur] == nxt) continue;
        p[nxt] = cur;
        depth[nxt] = depth[cur] + 1;
        DFS_2(nxt);
    }
}

File: Tree/tree_DFS.py
num_node = 7
edge = [(1,2), (1,3), (2,4), (2,5), (3,6), (3,7)]

adj = [[] for i in range(num_node+1)]
for i in range(len(edge)):
    u, v = edge[i]
    adj[u].append(v) 
    adj[v].append(u)
    
print("Tree adj list : {}".format(adj))

#vis = [False]*(num_node+1)
p = [0]*(num_node+1)        # 부모노드(반드시 0 삽입) & visited list 불필요
depth = [0]*(num_node+1)    # DFS : Tree 깊이 추적 가능

# Tree DFS 탐색 
def DFS(curr):
    print(curr, end = " ")
    for nxt in adj[curr]:
        if (p[curr] == nxt): continue
        p[nxt] = curr
        depth[nxt] = depth[curr] + 1
        DFS(nxt)
         

DFS(1)
print("\n")
print(p, end = "\n")
print(depth, end= "\n")
